src/mqt/ionshuttler/linear/dd/idealized_hahn.py

```
@@ -0,0 +1,227 @@
+    def from_dict(cls, data: object) -> IdealizedHahnReport:
+        """Restore an idealized-Hahn report from JSON-compatible values.
+
+        Returns:
+            The restored report.
+
+        Raises:
+            ValueError: If the serialized report is malformed.
+        """
+        if not isinstance(data, dict) or not isinstance(data.get("sequences"), list):
+            msg = "idealized Hahn report must contain a sequence list"
+            raise ValueError(msg)  # ruff: ignore[type-check-without-type-error] - Malformed JSON uses ValueError.
```

Ruff (0.16.1)

[warning] Prefer `TypeError` exception for invalid type

(TRY004)